packages/playwright/src/runner/testServer.ts

```diff
@@ -0,0 +1,125 @@
+      ws.on('message', async message => {
+        const { id, method, params } = JSON.parse(message.toString());
+        const result = await (dispatcher as any)[method](params);
+        ws.send(JSON.stringify({ id, result }));
+      });
+      return {
+        async close() {}
+      };
+    },
+  });
+  const url = await wss.listen(0, 'localhost', '/' + createGuid());
+  // eslint-disable-next-line no-console
+  console.log(`Listening on ${url}`);
+}
+
+class Dispatcher {
+  private _config: FullConfigInternal;
+
+  constructor(config: FullConfigInternal) {
+    this._config = config;
+  }
+
+  async test(params: { mode: 'list' | 'run', locations: string[], options: PlaywrightTestOptions, reporter: string, env: NodeJS.ProcessEnv }) {
+    for (const name in params.env)
+      process.env[name] = params.env[name];
+    if (params.mode === 'list')
+      await listTests(this._config, params.reporter, params.locations);
+    if (params.mode === 'run')
+      await runTests(this._config, params.reporter, params.locations, params.options);
+  }
+}
+
+
+async function listTests(config: FullConfigInternal, reporterPath: string, locations: string[] | undefined) {
+  config.cliArgs = [...(locations || []), '--reporter=null'];
+  const reporter = new InternalReporter(new Multiplexer(await createReporters(config, 'list', [[reporterPath]])));
+  const taskRunner = createTaskRunnerForList(config, reporter, 'out-of-process', { failOnLoadErrors: false });
+  const testRun = new TestRun(config, reporter);
+  reporter.onConfigure(config.config);
+
+  const taskStatus = await taskRunner.run(testRun, 0);
+  let status: FullResult['status'] = testRun.failureTracker.result();
+  if (status === 'passed' && taskStatus !== 'passed')
+    status = taskStatus;
+  const modifiedResult = await reporter.onEnd({ status });
+  if (modifiedResult && modifiedResult.status)
+    status = modifiedResult.status;
+  await reporter.onExit();
+}
+
+async function runTests(config: FullConfigInternal, reporterPath: string, locations: string[] | undefined, options: PlaywrightTestOptions) {
+  config.cliArgs = locations || [];
+  config.cliGrep = options.grep;
+  config.cliProjectFilter = options.projects;
+
+  config.configCLIOverrides.reporter = [[reporterPath]];
+  config.configCLIOverrides.repeatEach = 1;
+  config.configCLIOverrides.retries = 0;
+  config.configCLIOverrides.workers = options.oneWorker ? 1 : undefined;
+  config.configCLIOverrides.preserveOutputDir = true;
+  config.configCLIOverrides.use =  {
+    trace: options.trace,
+    headless: options.headed ? false : undefined,
+    _optionContextReuseMode: options.reuseContext ? 'when-possible' : undefined,
+    _optionConnectOptions: options.connectWsEndpoint ? { wsEndpoint: options.connectWsEndpoint } : undefined,
+  };
+
+  const reporter = new InternalReporter(new Multiplexer(await createReporters(config, 'run')));
+  const taskRunner = createTaskRunnerForWatch(config, reporter);
```

Comment:
This does not include global setup tasks. Did you mean to run them at the start of the dispatcher?